Introduce accurate min fee calculation function: `calcMinFeeTx`

lehins opened the pull request:
# Description

Addition of `calcMinFeeTx` function is a better alternative to `estimateMinFeeTx` introduced in #3962 

In order to implement this function it was necessary to:

* add `genesisKeyHashCount`, that counts up how many genesis key transaction will add
* Properly fix `getWitsVKeyNeeded` to also include `reqSignerHashes` as it was designed in the spec.

This last change is important, because that is the only change in this PR that actually affects the rules. Currently for Babbage and Alonzo it introduces a redundant check. It is not a problem since checking presence of a few extra hashes from `reqSignerHashes` in a map of provided hashes is not expensive at all. However this complete functionality of `getWitsVKeyNeeded` according to the spec and allows us to use it for fee calculation. Also when we enter Conway, we'l be able to get rid of the redundant check, see #3972

Moreover, this new function `calcMinFeeTx` was utilized in the ImpTest in order to take care fixing up the fee during creation of test transactions. Fixes #3767 

Some extra cleanup was also added:
* Deprecation of poorly named `txid` in favor of better named `txIdTxBody` and `txIdTx`
* Deprecation of poorly named `txup` function and switching of `PPUP` rule signal from `Maybe Update` to `StrinctMaybe Update` in order to reduce unnecessary conversions
* Add `ProtVerAtMost era 8` as a superclass constraint to `ShelleyEraTxBody` effectively disabling the type class from Conway onwards. This resulted in a cleanup of redundant constraints and fixup of over-constrained functions
* Also, `estimateMinFeeTx` functionality that was introduced in #3962 was moved to a new module `Cardano.Ledger.Tools` in order to make it accessible to the test suite

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

lehins committed:
Move `setMinFeeTx` and `estimateMinFeeTx`

Those functions can be useful in testing as well as for downstream
users, therefore it makes sense to start a new module where we can
collect functionality of this sorts.

lehins committed:
Prevent instances for `ShelleyEraTxBody` starting with protocol version 9.

This is a nice cleanup that allows us removing `ShelleyEraTxBody`
instance for `ConwayEra` onwards.

This commit also fixes a bunch of overconstraint functions and removes
some redundant constraints

lehins committed:
Change `PPUP` signal to `StrictMaybe` from `Maybe`

This allows us to deprecate the horribly named and useless `txup` function.

Also it simplifies the implementation a little bit by avoiding redundant
conversion

lehins committed:
Add `getAlonzoWitsVKeyNeeded` and use it for `getWitsVKeyNeeded`

Using `getAlonzoWitsVKeyNeeded` for Alonzo and Babbage is safe, because
it will check `reqSignerHashes` twice, but it allows us to rely on
`getWitsVKeyNeeded` to produce all of the required witnesses, aside from
the ones that are used in native scripts. This also makes implementation
match the spec.

https://github.com/IntersectMBO/cardano-ledger/issues/3972 tracks the
removal of `MissingRequiredSigners`

lehins committed:
Add `calcMinFeeTx`

lehins committed:
Add `getGenesisKeyHashCountTxBody`

* And `getShelleyGenesisKeyHashCountTxBody`

lehins committed:
Use `getGenesisKeyHashCount` for fee calculation

lehins committed:
Use new `calcMinFeeTx` in ImpTest

* Deprecate `txin`
* Add `txIdTx` and `txIdTxBody`
* Organize `submitTx` functions in ImpTest

lehins committed:
Add `txInsFilter`

lehins committed:
Add `lookupImpRootTxOut` and remove `impRootTxCoin`

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/Tools.hs:
```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE TypeOperators #-}
+
+-- | This is a module that contains functionality that is not necessary for ledger
+-- operation, but is usefuk for testing as well as for downstream users of ledger
```
I assume this should say `useful` ? :laughing:

lehins replied:
I think so too :laughing:

teodanciu commented on libs/cardano-ledger-core/CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## 1.10.0.0
 
+* Add `estimateMinFeeTx`, `addDummyWitsTx` and move `setMinFeeTx` to `Cardano.Ledger.Tools`
```
I wonder if we should mention the removal of `estimateMinFeeTx` and `addDummyWitsTx` functions in the cardano-ledger-api changelog as well, since maybe someone using the api might not know to look into core to figure out what happened?

lehins replied:
`estimateMinFeeTx` is still re-exported in `cardano-ledger-api`.

Also, neither of those two functions are in any release on chaps yet.

teodanciu commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/AdaPots.hs:
```diff
@@ -140,7 +140,7 @@ instance Show Produced where
 -- | Compute the Coin part of what is consumed by a TxBody, itemized as a 'Consume'
 consumedTxBody ::
-  ShelleyEraTxBody era =>
+  EraTxBody era =>
```
Just curious, I can't see the link between the main change  (adding `ProtVerAtMost era 8` to  `ShelleyEraTxBody` ) and this - was this maybe overconstrained as it was?

lehins replied:
Yes. It was just overconstraint.

teodanciu commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/TxBody.hs:
```diff
@@ -84,7 +84,7 @@ import GHC.Generics (Generic)
 import NoThunks.Class (NoThunks (..))
 
-class ShelleyEraTxBody era => AllegraEraTxBody era where
+class EraTxBody era => AllegraEraTxBody era where
```
I suppose removing `ShelleyEraTxBody`  constraint was necessary so that Conway can still have an instance for this, correct?

lehins replied:
Correct.

teodanciu commented on libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Body.hs:
```diff
@@ -89,10 +89,28 @@ evaluateTransactionBalance pp dpstate (UTxO u) txBody = consumed <-> produced
+  UTxO era ->
+  TxBody era ->
+  Value era
+evaluateTransactionBalanceShelley pp dpstate (UTxO u) txBody = consumed <-> produced
```
This duplication is a bit sad, couldn't we define the function  `evaluateTransactionBalance` above using this one: 
```
evaluateTransactionBalance pp dpstate utxo txBody =
    evaluateTransactionBalanceShelley pp dpstate utxo txBody<> (txBody ^. mintValueTxBodyF)
```
?

lehins replied:
Yes, we can. Thank you!!!

teodanciu commented on eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxow.hs:
```diff
@@ -312,10 +312,15 @@ babbageUtxowTransition = do
       witsVKeyNeeded must include the reqSignerHashes in the union   -}
   {- reqSignerHashes txbody ⊆ witsKeyHashes -}
+  -- TODO: Remove once we are in Conway. This check is now handled by
+  -- validateNeededWitnesses, as per spec
```
Won't it fail with a different Predicate failure in  `validateNeededWitnesses`  (`MissingVKeyWitnessesUTXOW`)  ?  Is that not something we need to be careful about changing?

lehins replied:
That's a very good question.

There are two important points in ledger that we need to consider with respect to predicate failures:

* If a transaction was invalid for any reason, it should continue being invalid in the future. Regardless of the type of predicate failure produced. That is necessary for consensus. In other words consensus does not care why transaction is invalid all it cares about is that it is invalid, because it doesn't propagate invalid transactions.
* We cannot start producing new types of predicate failures for current era that were not available before without a protocol version bump, because that would break NodeToClient communication protocol on the serialization level.

Neither one of those restrictions are preventing us from producing an extra predicate failure, which is what happening right now in this PR. In the matter of fact, writing this explanation up makes me realize that we can even produce a different predicate failure for the same predicate check, as long as that new type of predicate check existed before. So, technically we can take care of this TODO even now.

What we cannot do until we are in Conway, is to remove the actual `MissingRequiredSigners` type constructor, because that would break NodeToClient communication on the client side. This is because a newer client can be talking to an older node, that is capable producing this predicate failure.

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/Tools.hs:
```diff
@@ -47,6 +55,41 @@ setMinFeeTx pp tx =
+  -- actual witnesses supplied and the time when the transaction will be
+  -- submitted. Therefore we put this burden on the user.
+  --
+  -- This number can also be used to specify all of the redundand extra key witnesses that
```
typo: `redundant`

lehins replied:
Done

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/Core.hs:
```diff
@@ -235,6 +235,11 @@ class
 
+  -- | This function is not used in the ledger rules. It is only used by the downstream
+  -- tooling to figure out how many witnesses should be supplied for Genesis keys.
+  genesisKeyHashCount :: TxBody era -> Int
```
Do we have a convention for when to use functions prefixed by `get` in these typeclasses? I have seen a few cases where the version that didn't start with `get` was deprecate din favour of the `get`-prefixed one  (for example `minFee` and `getMinFree`). 
Is it perhaps to do with the comment that you added that it's not used in the rules?

lehins replied:
`minfee` used to be a standalone function. I was adding a type family suffix and prefixing with `get` every getter function that became a type class function, i.e. became era agnostic, like `getMinFeeTx` for example.

I think for consistency sake this one should be also named `getGenesisKeyHashCountTxBody`. 

That being said, there was no official convention enforced for naming. Which I think we should get together at some point and develop as a team.

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/Core.hs:
```diff
@@ -235,6 +235,11 @@ class
+  -- | This function is not used in the ledger rules. It is only used by the downstream
+  -- tooling to figure out how many witnesses should be supplied for Genesis keys.
+  genesisKeyHashCount :: TxBody era -> Int
+  genesisKeyHashCount _ = 0
```
Why have a default implementation - is it because they don't exist starting with Conway?

lehins replied:
Precisely. All future eras after Babbage will have it set to 0, therefore it is a safe default.

teodanciu commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/EpochSpec.hs:
```diff
@@ -38,7 +38,7 @@ spec = describe "EPOCH" $ do
       deposited <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosDepositedL
       deposited `shouldBe` zero
-    _ <- submitTx "simple transaction" $ mkBasicTx mkBasicTxBody
+    _ <- submitTxAnn "simple transaction" $ mkBasicTx mkBasicTxBody
```
You could use `submitTxAnn_` here

lehins replied:
Nice. Thank you. Done

teodanciu commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -470,56 +479,61 @@ fixupFees tx = do
+    outsWithoutRemaining = tx ^. bodyTxL . outputsTxBodyL
+    fee = calcMinFeeTx utxo pp txNoWits nativeScriptKeyWitsCount
+    remainderAvailable = remainingTxOut ^. coinTxOutL <-> fee
+    remainingTxOut' = remainingTxOut & coinTxOutL .~ remainderAvailable
```
We could define this as 
`remainingTxOut & coinTxOutL %~ (<-> fee)`, but maybe you wanted to emphasize the relationship to `remainderAvailable` ?

lehins replied:
`remainderAvailable` is used below. So that was the reason it is not collapses into the single modification using `%~`

teodanciu commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -470,56 +479,61 @@ fixupFees tx = do
-  wits <- mkTxWits (balancedTx ^. bodyTxL)
-  pure $ balancedTx & witsTxL .~ wits
+    nativeScriptKeyWitsCount = 0 -- TODO figure out wits for native scripts
+    outsWithoutRemaining = tx ^. bodyTxL . outputsTxBodyL
```
I don't understand why this is called `outsWithoutRemaining`,  it seems to be set to all the outs via `outputsTxBodyL` ?

Soupstraw replied:
I think the idea is that the leftover coin gets added to this later so it's not the final sequence of outputs.

lehins replied:
Yes, "remaining" is in the sense of "remainder of the money" or "change given back to you"

That was the naming Joosep adopted, so I just went with it as well

teodanciu replied:
Makes sense, thank you both :+1:

teodanciu commented on eras/shelley/impl/CHANGELOG.md:
```diff
@@ -5,7 +5,7 @@
 * Change `PPUP` signal to `StrictMaybe` from `Maybe`
 * Deprecate `txup` function.
-* Prevent instances for `ShelleyEraTxBody` starting with protocol version 9.
+* Prevent instances for `SahelleyEraTxBody` starting with protocol version 9.
```
Introduced a spurious `a` here in `ShelleyEraTxBody` :)

lehins replied:
Thank you. Fixed

teodanciu commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -726,10 +716,15 @@ addKeyPair kp@(KeyPair vk _) = do
+      error $
+        "Could not find a keypair corresponding to keyHash:"
+          ++ show keyHash
+          ++ "\nAlways use `freshKeyHash` to create key hashes."
```
Nice that it gives a hint about what could fix the problem :pray:

teodanciu reviewed (approved):
Looks good to me, tons of improvements packed in this PR :+1:

Soupstraw commented:
I assume it also closes #3970 (which is basically a duplicate of #3767)

Soupstraw commented on eras/babbage/impl/CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## 1.6.0.0
 
+* Use `getAlonzoWitsVKeyNeeded` for `getWitsVKeyNeeded`
```
Is it necessary to add this to the changelog? I thought we only have to add things that have changed at the type level.

lehins replied:
Also it is good to record any semantic changes to the behavior of a function.

In this case however, you are right that we don't need to add this line because `getWitsVKeyNeeded` was not even available in the previous version. 

So, thank you for pointing it out. I'll remove it.

Soupstraw commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/UTxO.hs:
```diff
@@ -233,3 +235,15 @@ getAlonzoScriptsNeeded (UTxO u) txBody =
     !minted = map (\pId@(PolicyID hash) -> (Minting pId, hash)) $ Set.toList $ txBody ^. mintedTxBodyF
+
+-- | Just like `getShelleyWitsVKeyNeeded`, but also requires `reqSignerHashesTxBodyL`.
+getAlonzoWitsVKeyNeeded ::
```
I wonder if it'd be a good idea to add a `ProtVerAtMost` constraint to `get*WitsVKeyNeeded` functions. That would make it harder to accidentally use e.g. `getShelleyWitsVKeyNeeded` in Babbage when we have a newer version of this function available in Alzono.

lehins replied:
But this function actually uses `getShelleyWitsVKeyNeeded`.

Soupstraw replied:
oh, right, never mind, that would be too much hassle

Soupstraw commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -450,76 +456,90 @@ getRewardAccountAmount rewardAcount = do
-        (State (EraRule "LEDGER" era), Tx era)
-    )
-submitTx_ tx = do
+  ImpTestM era (Either [PredicateFailure (EraRule "LEDGER" era)] (Tx era))
```
I think I'd prefer returning the `TxId` rather than the `Tx` in all of the ImpTest functions that deal with submitting a transaction. Returning a `Tx` would also work since you can access the `TxId` from it as well, but it's a good idea to agree on which one we're going to use and then use that across all `ImpTest` modules. 

The way I'd like to think about it is that `Tx` is used for transactions that are not yet submitted to the chain and `TxId` represents a transaction on the chain. That would have the added bonus that whenever you submit a transaction, you'll actually have to look it up from the UTxO and we might catch some additional bugs with that.

lehins replied:
I totally hear you. However, the problem is that transaction that was supplied to the function is not the transaction that gets submitted. We modify it by fixing up the fee and by adding witnesses to it. Without returning the modified transaction there is no way to get access to what has been submitted, which can be useful for some tests.

That was the main reason why I switched from the `TxId` to `Tx`.

Soupstraw replied:
Oh right, we can't get much besides the UTxO entry with the `TxId`, yeah that makes sense.

Soupstraw commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -329,22 +331,23 @@ trySubmitVote ::
-              )
-          )
+trySubmitVote vote voter gaId =
+  fmap (fmap txIdTx) $
```
We wouldn't have to do this if `trySubmitTx` returned a `TxId`, see my comment above

lehins replied:
See my other comment.

If it proves to be a common enough pattern then we can create extra helper functions that return TxId instead of Tx, but calling `txIdTx` on the result is quite easy.

Soupstraw reviewed (commented):
Looks good overall. 

I have bit of mixed feelings about losing the `ShelleyEraTxBody` superclass, since it makes the system of typeclasses less uniform. Sure, we can get rid of some `ProtVerAtMost` constraints, but we also have to add an explicit `ShelleyEraTxBody` constraint to some functions that didn't need it before. Then again it's nice to not have to implement `ShelleyEraTxBody` with a bunch of dummy methods from Conway onwards.

lehins commented:
> I have bit of mixed feelings about losing the ShelleyEraTxBody superclass, since it makes the system of typeclasses less uniform. 

It allows us to deprecate a lot of functionality in bulk. Starting with the next era that follows  Conway we'll do the same thing for `ShelleyEraTxCert`. So, it will become a uniform patter over time: if all functionality from the era is deprecated we can deprecate the whole class.

>  we can get rid of some ProtVerAtMost constraints, we also have to add an explicit ShelleyEraTxBody constraint to some functions that didn't need it before

We can replace one for another. That's all that has happend in this PR.

> Then again it's nice to not have to implement ShelleyEraTxBody with a bunch of dummy methods from Conway onwards.

Yeah for all of the eras that will follow.

I think it makes the deprecation process much cleaner when we can deprecate a class as a whole.

lehins committed:
Deprecate `requiredSignersAreWitnessed`

We can now remove the duplicated check for `reqSignerHashes`, however we
still need to keep around the predicate failure for NodeToClient
comunication.

lehins committed:
Fixup
